DMCompiler/Compiler/DMPreprocessor/DMPreprocessorParser.cs

```diff
@@ -0,0 +1,294 @@
+﻿using DMCompiler.Compiler.DM;
+using DMCompiler.DM;
+using OpenDreamShared.Compiler;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+#nullable enable
+namespace DMCompiler.Compiler.DMPreprocessor {
+    /*
+     * NOTE: Some operations are (as in BYOND) not implemented for preproc expressions.
+     * This includes ternary, bitwise, and assignment.
+    Grammar: (As usual in extended BNF, {A} means 0 or more As, and [A] means an optional A. Brace characters are escaped with single-quotes.)
+    Expression ::= AndExp ['||' AndExp]
+    AndExp ::= CompExp ['&&' CompExp]
+    CompExp ::= LtGtExp [('==' | '!=') LtGtExp]
+    LtGtExp ::= AddSubExp [(‘<’ | ‘<=’ | ‘>’ | ‘>=’ | ‘==’ | ‘!=’) AddSubExp]
+    AddSubExp ::= MultExp [('+' | '-') MultExp]
+    MultExp ::= PowerExp [('*' | '/' | '%') PowerExp]
+    PowerExp ::= UnaryExp [** PowerExp]
+    UnaryExp ::= {!}SignExp
+    SignExp ::= [-]Primary
+    Primary ::= '(' Expression ')' | DefinedExp | Constant
+    DefinedExp ::= 'defined(' MacroIdentifier ')'
+    Constant ::= Integer | Float | Null
+    */
+
+    /// <summary>
+    /// An extremely simple parser that acts on Preproc tokens, <br/>
+    /// held separate from DMParser because of slightly different behaviour and far simpler implementation.
+    /// </summary>
+    internal static class DMPreprocessorParser {
+        private static List<Token>? _tokens;
+        private static Dictionary<string, DMMacro>? _defines;
+        private static int _tokenIndex = 0;
+        private static readonly float DegenerateValue = 0.0f;
+
+        /// <returns>A float, because that is the only possible thing a well-formed preproc expression can evaluate to.</returns>
+        public static float? ExpressionFromTokens(List<Token> input, Dictionary<string,DMMacro> defines) {
+            _tokens = input;
+            _defines = defines;
+            var ret = Expression();
+            _tokens = null;
+            _defines = null;
+            _tokenIndex = 0;
+            return ret;
+        }
+
+        private static void Advance() {
+            ++_tokenIndex;
+        }
+        private static Token? Current() {
+            if (_tokenIndex >= _tokens!.Count())
+                return null;
+            return _tokens![_tokenIndex];
+        }
+        private static bool Check(TokenType type) {
+            if (Current()?.Type == type) {
+                Advance();
+                return true;
+            }
+            return false;
+        }
+        private static bool Check(TokenType[] types) {
+            foreach (TokenType type in types) {
+                if (Current()?.Type == type) {
+                    Advance();
+                    return true;
+                }
+            }
+            return false;
+        }
+        private static void Error(string msg) {
+            DMCompiler.Error(new CompilerError(Current()?.Location ?? Location.Unknown, msg));
```

Comment:
Use the `Error(Location loc, string message)` version instead.
```suggestion
            DMCompiler.Error(Current()?.Location ?? Location.Unknown, msg);
```